# ChatGPT Mental Support simulation

In this notebook we will simulate how chatGPT will respond to request from people with mental issues, problems and worries, that will be expressed plain in conversation.
This responses will be compared to responses from mental counselor to see how is chatGPT doing this.

### Dataset

This simulation is based on this dataset:

[Mental health Counseling Conversations dataset](https://www.kaggle.com/datasets/thedevastator/nlp-mental-health-conversations)

### Dataset structure

| Column   | Description | 
|----------|----------|
| Context  | Request, phrase, query sentence from the patient, that seems to have mental troubles, problem, disorders of different kind        |  
| Response | Response, that mental counselor is supposed to give. For one ```Context``` there is from 5 to 25 different responses | 


### Comparison

Comparison between ChatGPT response will be done with special prompt, resulting in score [0-100], and be done comparing one response from ChatGRT to all (one by one) responses from the dataset. New dataset will chatGPT responses and scores will be build, and then explored



## Preparing the data

The dataset on the link above is available to download, so here it is already present in the ```dataset/NLP_Mental_Health_Conversations.csv```

In [86]:
#!pip install pandas

In [87]:
import pandas as pd

In [88]:
CONVERSATIONS_DATASET_FILENAME='dataset/NLP_Mental_Health_Conversations.csv'

conversation_dataset=pd.read_csv(CONVERSATIONS_DATASET_FILENAME)

conversation_dataset.head()

Context  \
0  I'm going through some things with my feelings...   
1  I'm going through some things with my feelings...   
2  I'm going through some things with my feelings...   
3  I'm going through some things with my feelings...   
4  I'm going through some things with my feelings...   

                                            Response  
0  If everyone thinks you're worthless, then mayb...  
1  Hello, and thank you for your question and see...  
2  First thing I'd suggest is getting the sleep y...  
3  Therapy is essential for those that are feelin...  
4  I first want to let you know that you are not ...

In [89]:
len(conversation_dataset)

3512

In [90]:
conversation_dataset.columns = conversation_dataset.columns.str.lower().str.replace(' ', '_')

In [91]:
conversation_dataset.isna().sum()

context     0
response    4
dtype: int64

In [92]:
conversation_dataset.describe()

context  \
count                                                3512   
unique                                                995   
top     I have so many issues to address. I have a his...   
freq                                                   94   

                                                 response  
count                                                3508  
unique                                               2479  
top     It's normal to feel a little anxiety--after al...  
freq                                                    3

In [93]:
context_example=conversation_dataset.loc[328]['context']

responses_example=conversation_dataset[conversation_dataset['context']==context_example]

print(context_example)


I'm in a state of depression right now. Who can I talk to? I've been sick in a lot of pain and crying. Don't know where to turn.


In [94]:
responses_example

context  \
328   I'm in a state of depression right now. Who ca...   
2166  I'm in a state of depression right now. Who ca...   

                                               response  
328   First of all, if you’re sick and in pain, have...  
2166  First of all, if you’re sick and in pain, have...

In [95]:
responses_example.reset_index(inplace=True)
responses_example
print(responses_example.loc[0]['response'])
print(responses_example.loc[1]['response'])

First of all, if you’re sick and in pain, have you seen your doctor? If you have a chronic illness and pain, this can often lead to depression and is very common. You’ve taken the first step to dealing with it. You can talk to your doctor, even if your depression isn’t related to your health. He or she can either look at prescribing you an antidepressant or can refer you to a psychologist or psychiatrist.If that’s not really an option for you, there are also helplines (local or national) or support groups where people dealing with similar issues can talk to each other. If you can't find any in your area (newspapers usually publish this information), there are online support groups as well.
First of all, if you’re sick and in pain, have you seen your doctor? If you have a chronic illness and pain, this can often lead to depression and is very common. You’ve taken the first step to dealing with it. You can talk to your doctor, even if your depression isn’t related to your health. He or s

### Dataset contains duplicate responses

lets see some more

In [96]:
context_example=conversation_dataset.loc[38]['context']

responses_example=conversation_dataset[conversation_dataset['context']==context_example]

print(context_example)

I have so many issues to address. I have a history of sexual abuse, I’m a breast cancer survivor and I am a lifetime insomniac.    I have a long history of depression and I’m beginning to have anxiety. I have low self esteem but I’ve been happily married for almost 35 years.
   I’ve never had counseling about any of this. Do I have too many issues to address in counseling?


In [97]:
responses_example

context  \
23    I have so many issues to address. I have a his...   
24    I have so many issues to address. I have a his...   
25    I have so many issues to address. I have a his...   
26    I have so many issues to address. I have a his...   
27    I have so many issues to address. I have a his...   
...                                                 ...   
2010  I have so many issues to address. I have a his...   
2011  I have so many issues to address. I have a his...   
2012  I have so many issues to address. I have a his...   
2013  I have so many issues to address. I have a his...   
2014  I have so many issues to address. I have a his...   

                                               response  
23    Let me start by saying there are never too man...  
24    It is never too late to get help and begin mak...  
25    You have been through so much and it sounds li...  
26    Absolutely not.  I strongly recommending worki...  
27    Absolutely not!  In fact, most people have man...  
...                                                 ...  
2010  The way it goes is that all your issues are co...  
2011  I think this is a very common question that pe...  
2012  Hello, I'm so glad you decided to take the fir...  
2013  Not at all my dear. Human beings are complex c...  
2014  Everyone has topics to discuss when they start...  

[94 rows x 2 columns]

In [98]:
context_example=conversation_dataset.loc[1000]['context']

responses_example=conversation_dataset[conversation_dataset['context']==context_example]

print(context_example)

I'm fine when we start becoming intimate, but out of nowhere, I will get a flashback of what happened to me in the past. I start hysterically crying and freaking out when my boyfriend obviously has done nothing to hurt me.


In [99]:
responses_example

context  \
999   I'm fine when we start becoming intimate, but ...   
1000  I'm fine when we start becoming intimate, but ...   
2701  I'm fine when we start becoming intimate, but ...   
2709  I'm fine when we start becoming intimate, but ...   

                                               response  
999   Have you sought the support of a therapist rea...  
1000  Sexual intimacy can be very triggering for sur...  
2701  Have you sought the support of a therapist rea...  
2709   Sexual intimacy can be very triggering for su...

### Removing duplicates and zero entries

In [100]:
conversation_dataset.dropna(subset=['response'], inplace=True)
conversation_dataset.drop_duplicates(subset=['response'], inplace=True)
conversation_dataset.reset_index(drop=True, inplace=True)

In [101]:
len(conversation_dataset)

2479

In [102]:
conversation_dataset.isna().sum()

context     0
response    0
dtype: int64

In [103]:
context_example=conversation_dataset.loc[328]['context']

responses_example=conversation_dataset[conversation_dataset['context']==context_example]

print(context_example)


I'm in a state of depression right now. Who can I talk to? I've been sick in a lot of pain and crying. Don't know where to turn.


In [104]:
responses_example

context  \
328   I'm in a state of depression right now. Who ca...   
2025  I'm in a state of depression right now. Who ca...   

                                               response  
328   First of all, if you’re sick and in pain, have...  
2025  First of all, if you’re sick and in pain, have...

In [105]:
conversation_dataset.describe()

context  \
count                                                2479   
unique                                                912   
top     I have so many issues to address. I have a his...   
freq                                                   47   

                                                 response  
count                                                2479  
unique                                               2479  
top     If everyone thinks you're worthless, then mayb...  
freq                                                    1

In [106]:
context_example=conversation_dataset.loc[1000]['context']

responses_example=conversation_dataset[conversation_dataset['context']==context_example]

print(context_example)

I'm fine when we start becoming intimate, but out of nowhere, I will get a flashback of what happened to me in the past. I start hysterically crying and freaking out when my boyfriend obviously has done nothing to hurt me.


In [107]:
responses_example

context  \
999   I'm fine when we start becoming intimate, but ...   
1000  I'm fine when we start becoming intimate, but ...   
2221  I'm fine when we start becoming intimate, but ...   
2226  I'm fine when we start becoming intimate, but ...   

                                               response  
999   Have you sought the support of a therapist rea...  
1000  Sexual intimacy can be very triggering for sur...  
2221  Have you sought the support of a therapist rea...  
2226   Sexual intimacy can be very triggering for su...

In [108]:
conversation_dataset.loc[1000]['response']==conversation_dataset.loc[2226]['response']

False

In [109]:
print(conversation_dataset.loc[1000]['response'])
print(conversation_dataset.loc[2226]['response'])

print(conversation_dataset.loc[328]['response'])
print(conversation_dataset.loc[2025]['response'])

print(conversation_dataset.loc[328]['response']==conversation_dataset.loc[2025]['response'])

Sexual intimacy can be very triggering for survivors even when it is both wanted and consensual.  You may want to consider seeing a therapist who specializes in trauma to work through the abuse if you have not already done so. Often times triggers still hold such a powerful effect when the emotions related to the abuse  have not been fully processed.  In the meantime, you may want to consider coming up with a Safe Word to let your partner know that you are being triggered or to communicate your physical boundaries to him.  Often times, the experience of communicating  your physical boundaries to your partner, having those boundaries respected and validated, and having a partner who is understanding and  willing to engage in intimacy in such a way that does not violate your physical boundaries  can reinforce a sense of safety with him.
 Sexual intimacy can be very triggering for survivors even when it is both wanted and consensual.  You may want to consider seeing a therapist who specia

In [110]:
response1 = conversation_dataset.loc[328]['response']
response2 = conversation_dataset.loc[2025]['response']

# Find the minimum length of the two responses
min_length = min(len(response1), len(response2))

# Iterate through each character and find the positions where they differ
differences = []
for i in range(min_length):
    if response1[i] != response2[i]:
        differences.append(i)

# Print the positions where the responses differ
print("Positions where responses differ:")
print(differences)
print('Lengths of response1 and response2:', len(response1), len(response2))

Positions where responses differ:
[]
Lengths of response1 and response2: 698 699


In [111]:
print(response1.strip()==response2.strip())

True


So there is duplicate entries, but they differ in heading or trailing spaces.
Let's fix this, and remove duplicates once more.

In [112]:
conversation_dataset['response']=conversation_dataset['response'].apply(lambda x: x.strip())

In [113]:
conversation_dataset.describe()

context  \
count                                                2479   
unique                                                912   
top     I have so many issues to address. I have a his...   
freq                                                   47   

                                                 response  
count                                                2479  
unique                                               2075  
top     This can be a very complicated situation. The ...  
freq                                                    2

In [114]:
conversation_dataset.drop_duplicates(subset=['response'], inplace=True)
conversation_dataset.reset_index(drop=True, inplace=True)

In [115]:
conversation_dataset.describe()

context  \
count                                                2075   
unique                                                847   
top     I have so many issues to address. I have a his...   
freq                                                   47   

                                                 response  
count                                                2075  
unique                                               2075  
top     If everyone thinks you're worthless, then mayb...  
freq                                                    1

In [116]:
# Test once more
context_example=conversation_dataset.loc[1000]['context']

responses_example=conversation_dataset[conversation_dataset['context']==context_example]

print(context_example)

I'm fine when we start becoming intimate, but out of nowhere, I will get a flashback of what happened to me in the past. I start hysterically crying and freaking out when my boyfriend obviously has done nothing to hurt me.


In [117]:
responses_example

context  \
999   I'm fine when we start becoming intimate, but ...   
1000  I'm fine when we start becoming intimate, but ...   

                                               response  
999   Have you sought the support of a therapist rea...  
1000  Sexual intimacy can be very triggering for sur...

### Errors in dataset

Working with the dataset I found some irrelevant entries. So we should just delete them


In [136]:
error_context=[" such as not enough sleep", 
"Can a counselor take sides with one parent and allow a parent to order the child to tell the counselor"]

conversation_dataset=conversation_dataset[~conversation_dataset['context'].str.startswith(tuple(error_context))]


In [137]:
conversation_dataset.describe()

context  \
count                                                2072   
unique                                                845   
top     I have so many issues to address. I have a his...   
freq                                                   47   

                                                 response  
count                                                2072  
unique                                               2072  
top     If everyone thinks you're worthless, then mayb...  
freq                                                    1

In [138]:
context_unique = conversation_dataset.groupby('context').size().reset_index(name='count')

In [139]:
context_unique.sort_values('count',ascending=False)

context  count
286  I have so many issues to address. I have a his...     47
78   How do you know you have the right therapist f...     41
546  Is it normal for people to cry during therapy,...     38
524  I've gone to a couple therapy sessions so far ...     26
462  I'm going through some things with my feelings...     23
..                                                 ...    ...
361  I stress over everything. If I don't have enou...      1
366  I think I have depression, anxiety, bipolar di...      1
369  I think adult life is making him depressed and...      1
370  I think i may suffer from depression, and it i...      1
844  i was addicted to porn since the second grade....      1

[845 rows x 2 columns]

In [140]:
context_unique.sort_values('count',ascending=False).head(30)

context  count
286  I have so many issues to address. I have a his...     47
78   How do you know you have the right therapist f...     41
546  Is it normal for people to cry during therapy,...     38
524  I've gone to a couple therapy sessions so far ...     26
462  I'm going through some things with my feelings...     23
832  When my daughter is stressed about a silly thi...     23
80     How does a person start the counseling process?     22
31   Does counseling really do anything that can he...     21
79   How does a counselor decide when to end counse...     21
73   How do I decide if I should end my relationshi...     15
41   Every winter I find myself getting sad because...     15
699  My mother takes care of niece whom my sister a...     13
354  I start counseling/therapy in a few days (I'm ...     12
305  I just took a job that requires me to travel f...     11
311  I keep having these random thoughts that I don...     11
647  My fiancé and I have been together for 3 years...     11
827  When I'm in large crowds I get angry and I jus...     11
397  I was raped a couple months ago, Since then, a...     11
697  My mother has Alzheimer's and she has become s...     10
213  I have a friend that who I used to be in a rel...     10
485  I'm obsessing about a terrible breakup. Everyt...      9
817                What makes a healthy marriage last?      9
486  I'm planning to have baby, so I have to quit s...      9
568  I’ve been in a long distance relationship with...      9
154  I don't know what to say. I have never really ...      9
626  My daughter is in later elementary school. She...      9
563  I’m facing severe depression and anxiety and I...      9
16   After he got home from the hospital he was ang...      9
303  I just lost my grandpa and i'm having a rough ...      9
444  I'm being verbally abused on a daily basis by ...      9

In [141]:
context_unique['count'].describe()

count    845.000000
mean       2.452071
std        3.549132
min        1.000000
25%        1.000000
50%        1.000000
75%        2.000000
max       47.000000
Name: count, dtype: float64

In [142]:
context_unique[context_unique['count']==1]

context  count
0    A few nights ago I talked to this girl I know ...      1
2    A friend of mine taking psychology advised I g...      1
3    A girl and I were madly in love. We dated for ...      1
8    About a month ago, I went through my boyfriend...      1
13   After a domestic abuse situation, I went to a ...      1
..                                                 ...    ...
830  When having sex I think of other men and want ...      1
837  Whenever I leave my girlfriend I get panic att...      1
839  Whenever I sing along with a song that I like ...      1
841                   Why am I attracted to older men?      1
844  i was addicted to porn since the second grade....      1

[452 rows x 2 columns]

In [145]:
count_distribution = context_unique['count'].value_counts()

count_distribution

count
1     452
2     184
3      74
4      43
5      25
6      18
9      12
7      11
8       6
11      5
23      2
21      2
15      2
10      2
38      1
26      1
41      1
12      1
47      1
22      1
13      1
Name: count, dtype: int64

In [144]:
conversation_dataset.describe()

context  \
count                                                2072   
unique                                                845   
top     I have so many issues to address. I have a his...   
freq                                                   47   

                                                 response  
count                                                2072  
unique                                               2072  
top     If everyone thinks you're worthless, then mayb...  
freq                                                    1

### Data is cleaned and ready for simulation

So we have 845 unique requests, with a mean of 2.45 responses for each, and 452 having one response, and max of 47 responses, but only 10 requests have more then 10 responses.

### Conclusions

**Maybe for simulation we should use only requests with 2 or more responses, so we have some variety in approved responses to compare with. If we take request with 2 or more we will have 393 requests, if we take with 3 or more - 209 requests. 209 may be enough for our simulation. And having only one response will make it possible for incorrect scoring of ChatGPT output. Because mental issues is delicate matter, so several different responses may serve well for specific case.**


## ChatGPT